patches/server/0642-Datapack-API.patch

```diff
@@ -0,0 +1,257 @@
++    public boolean isEnabled() {
++        return enabled;
++    }
++
++    @Override
++    public void setEnabled(boolean enabled) {
++        if (enabled == this.enabled) {
++            return;
++        }
++
++        MinecraftServer server = MinecraftServer.getServer();
++        List<String> enabledKeys = server.getPackRepository().getSelectedPacks().stream().map(Pack::getId).collect(Collectors.toList());
++        if (enabled) {
++            enabledKeys.add(this.name);
++        } else {
++            enabledKeys.remove(this.name);
++        }
++        server.reloadResources(enabledKeys, ServerResourcesReloadedEvent.Cause.PLUGIN);
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/datapack/PaperDatapackManager.java b/src/main/java/io/papermc/paper/datapack/PaperDatapackManager.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..131f29bb86aded2c01271bf8fbfd7adc74765355
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/datapack/PaperDatapackManager.java
+@@ -0,0 +1,68 @@
++package io.papermc.paper.datapack;
++
++import net.minecraft.server.packs.FilePackResources;
++import net.minecraft.server.packs.FolderPackResources;
++import net.minecraft.server.packs.PackResources;
++import net.minecraft.server.packs.ResourcePackFileNotFoundException;
++import net.minecraft.server.packs.repository.Pack;
++import net.minecraft.server.packs.repository.PackRepository;
++import net.minecraft.server.packs.repository.PackSource;
++import org.apache.commons.lang.Validate;
++import org.jetbrains.annotations.NotNull;
++
++import java.io.File;
++import java.io.FileNotFoundException;
++import java.io.IOException;
++import java.util.Collection;
++import java.util.function.Supplier;
++import java.util.stream.Collectors;
++
++public class PaperDatapackManager implements DatapackManager {
++    private final PackRepository repository;
++
++    public PaperDatapackManager(PackRepository repository) {
++        this.repository = repository;
++    }
++
++    @NotNull
++    @Override
++    public Collection<Datapack> getPacks() {
++        Collection<Pack> enabledPacks = repository.getSelectedPacks();
++        return repository.getAvailablePacks().stream().map(loader -> new PaperDatapack(loader, enabledPacks.contains(loader))).collect(Collectors.toList());
++    }
++
++    @NotNull
++    @Override
++    public Collection<Datapack> getEnabledPacks() {
++        return repository.getSelectedPacks().stream().map(loader -> new PaperDatapack(loader, true)).collect(Collectors.toList());
++    }
++
++    @Override
++    public Datapack loadDatapack(@NotNull String name, @NotNull File file, boolean enable) throws FileNotFoundException, IOException {
++        Validate.notNull(name, "Name cannot be null");
++        Validate.notNull(file, "File cannot be null");
++        Pack pack;
++        try {
++            pack = Pack.createUnsafe(name, false, createSupplier(file), this.repository.constructor, Pack.Position.TOP, PackSource.SERVER);
```

Comment:
Is TOP really the best option here? Unless I'm misunderstanding here, that means it will be loaded first after the reload versus last. I would think it should be loaded after all the other packs are reloaded.